### PR TITLE
KV Cache MLA NaN Write Reporting

### PR DESCRIPTION
**Pull request by elvircrn**
## Purpose

Background: 

We discovered that R1 NVFP4 had issue with NaNs in the forward pass. While debugging, https://github.com/vllm-project/vllm/pull/43318 was created to aid debugging resulting in us finding and fixing and/or reporting a multitude of NaN-inducing bugs such as:
* https://github.com/tlrmchlsmth/vllm/pull/33
* https://github.com/Dao-AILab/flash-attention/issues/1974
* https://github.com/vllm-project/vllm/pull/35219
* https://github.com/vllm-project/vllm/pull/39444
etc.

While useful, the changes introduced by the aforementioned PR https://github.com/vllm-project/vllm/pull/43318 were introduced to maximize debugging efficiency and as such cannot be merged into vllm. Therefore, I propose to split the PR up into multiple parts:
1. detect NaN writes to KV$ (this PR)
2. detect which kernel was responsible for the NaN write (follow-up)

Therefore, the purpose of this PR is to introduce NaN KV$ writes with a focus on causality. Specifically, we would like to know exactly which phase (prefill/decode), rank (currently only EP is considered) and layer introduced the first NaN into the KV$.

Additionally we have the following constraints:
1. full cudagraph support
2. cover the cases where we erroneously do KV$ writes such as https://github.com/vllm-project/vllm/pull/39444/changes
3. make the reporting lightweight to cover high concurrency decode scenarios such as https://github.com/deepseek-ai/DeepEP/issues/621
4. must be usable in a wide EP (i.e P8+/D48) scenario 
5. the PR should make the checker optional and when disabled, should not have any effect on the runtime

Extending `concat_and_cache_mla` with `constexpr` covers all of the above cases, as the value to be written into the KV$ is in the registers and the kernel itself is lightweight.

Pushing reports to prometheus allows easier production debugging and optional monitoring.

We can refer to `VLLM_COMPUTE_NANS_IN_LOGITS`  as a reference implementation already in `main` which tackles a similar problem.

## Test Plan

Two test files, both under `tests/kernels/`:

**`test_kv_cache_nan_debug.py`** (15 tests, ~3s, no model loading):
- Kernel-level: NaN detection in kv_c, k_pe, clean inputs, accumulation, reset, padded slots, auto dtype
- Env var gating: VLLM_DEBUG_MLA_CACHE on/off
- Phase tracking (unit): prefill/decode labeling, origin lock-in, timestamp recording
- Kernel-to-stats integration: two-layer prefill origin detection

```bash
VLLM_DEBUG_MLA_CACHE=1 python -m pytest tests/kernels/test_kv_cache_nan_debug.py -v
```

test_kv_cache_nan_e2e.py (4 tests, ~50s, marked slow_test):
- Prometheus counters populated with correct phase/layer/origin/timestamp labels
- Clean weights produce zero NaN counters
- CUDA graph (FULL mode): NaN counter survives graph capture and replay
- Multiprocessing (VLLM_ENABLE_V1_MULTIPROCESSING=1): NaN injection via sitecustomize.py, detection across process boundary

```
VLLM_DEBUG_MLA_CACHE=1 python -m pytest tests/kernels/test_kv_cache_nan_e2e.py -v
```

## Test Result

```
$ VLLM_DEBUG_MLA_CACHE=1 python -m pytest tests/kernels/test_kv_cache_nan_debug.py -v
15 passed in 2.73s
```

```
$ VLLM_DEBUG_MLA_CACHE=1 python -m pytest tests/kernels/test_kv_cache_nan_e2e.py -v
4 passed in 48.59s
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/metrics/loggers.py [hunk 1]
```diff
@@ -539,6 +540,23 @@ def __init__(
...
+            self._kv_nan_rank = (
+                str(dist.get_rank()) if dist.is_initialized() else "0"
+            )
+            self.counter_kv_cache_nans = self._counter_cls(
+                name="vllm:kv_cache_nans_total",
+                documentation="NaN values detected in MLA KV cache FP8 conversion.",
+                labelnames=["rank", "layer", "phase"],
+            )
+            self.gauge_kv_nan_first_seen = self._gauge_cls(
+                name="vllm:kv_cache_nan_first_seen_timestamp",
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py:563]: This is a pretty bizarre thing to store in a time-series database at first glance - i.e. in Prometheus, there will be a timestamp associated with every value of the nans counter

Are you doing this because higher precision is needed? What do users do with these timestamps?

File: vllm/envs.py [hunk 2]
```diff
@@ -1640,16 +1563,29 @@ def _resolve_rust_frontend_path() -> str | None:
     "VLLM_COMPUTE_NANS_IN_LOGITS": lambda: bool(
         int(os.getenv("VLLM_COMPUTE_NANS_IN_LOGITS", "0"))
     ),
+    "VLLM_DEBUG_MLA_CACHE": lambda: bool(
```
Comment (markmc) [inline vllm/envs.py]: Why not proper config for this? See #25700

File: vllm/v1/metrics/loggers.py [hunk 3]
```diff
@@ -1146,6 +1164,18 @@ def record(
...
+        if envs.VLLM_DEBUG_MLA_CACHE:
+            for (layer, phase), count in iteration_stats.kv_cache_nans.items():
+                self.counter_kv_cache_nans.labels(
+                    rank=self._kv_nan_rank, layer=layer, phase=phase,
+                ).inc(count)
+            for (layer, phase), ts in iteration_stats.kv_cache_nan_first_seen.items():
+                g = self.gauge_kv_nan_first_seen.labels(
+                    rank=self._kv_nan_rank, layer=layer, phase=phase,
+                )
+                cur = g._value.get()
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py]: I'd take the underscore here to mean it's a private API

File: vllm/v1/metrics/loggers.py [hunk 4]
```diff
@@ -539,6 +540,23 @@ def __init__(
                 counter_corrupted_requests, per_engine_labelvalues
             )
 
+        if envs.VLLM_DEBUG_MLA_CACHE:
+            import torch.distributed as dist
+            self._kv_nan_rank = (
+                str(dist.get_rank()) if dist.is_initialized() else "0"
```
Comment (markmc) [inline vllm/v1/metrics/loggers.py]: This code is called in the frontend? Have you just been testing with fully in-proc `LLMEngine`?

(Which makes it seem weird to use Prometheus at all, except of course ... it's the basis for our `get_metrics()` API!)

**Review by markmc — COMMENTED** [review]
I know this is only in draft, just some comments FWIW ...

The "first seen timestamp" thing makes me wonder do we want something more like a core-dump for this? If we detect a nan, dump a bunch of information to a timestamped file for debugging?

**Comment by elvircrn** [comment]
@markmc Updated PR description, adding context.

File: vllm/v1/metrics/loggers.py (hunk 1 above)
Comment (elvircrn) [reply vllm/v1/metrics/loggers.py:563]: It's not bizzare because we want the most precise time recording of this event at the time of recording, not at the time of publishing. 

We don't want to loose causality info, the most important goal of this PR.

File: vllm/envs.py (hunk 2 above)
Comment (elvircrn) [reply vllm/envs.py]: We can do that once the PR is finalized.

File: vllm/v1/metrics/loggers.py (hunk 3 above)
Comment (elvircrn) [reply vllm/v1/metrics/loggers.py]: Noted!

File: vllm/v1/metrics/loggers.py (hunk 4 above)
Comment (elvircrn) [reply vllm/v1/metrics/loggers.py]: Been testing with `VLLM_ENABLE_V1_MULTIPROCESSING=0` so far but made an attempt at doing a similar solution that `COMPUTE_NAN_IN_LOGITS=1` implemented (as per the subsequently added PR description).

But the `torch.distributed.get_rank()` seems wrong there, so updated the code to use `vllm_config.parallel_config.data_parallel_rank` instead.

**Comment by elvircrn** [comment]
> I know this is only in draft, just some comments FWIW ...
> 
> The "first seen timestamp" thing makes me wonder do we want something more like a core-dump for this? If we detect a nan, dump a bunch of information to a timestamped file for debugging?

@markmc  We can do both, we can dump whatever we find useful at the Sampler. I think it makes sense to do that in this PR.

An additional idea for a follow-up PR would be to dump all tensors going into `concat_and_cache` when we encounter a NaN in a cudagraph-friendly way, ensuring that we only dump the tensors once during the first NaN occurence downstream in the Sampler.

File: csrc/libtorch_stable/cache_kernels.cu [hunk 5]
```diff
@@ -388,7 +388,36 @@ __global__ void reshape_and_cache_flash_kernel(
...
-template <typename scalar_t, typename cache_t, Fp8KVCacheDataType kv_dt>
+// nan_flag bit layout (FP8 KV cache only):
+//   bit 0 = FP8 NaN in kv_c write    (byte & 0x7F) == 0x7F
+//   bit 1 = FP8 NaN in k_pe write
+//   bit 2 = Inf in kv_c source
+//   bit 3 = Inf in k_pe source
+//   bit 4 = NaN in kv_c source
+//   bit 5 = NaN in k_pe source
+
+__device__ __forceinline__ bool is_inf_bits(__nv_bfloat16 v) {
```
Comment (elvircrn) [inline csrc/libtorch_stable/cache_kernels.cu]: Note to self: move to utils.

File: serve_nan_debug.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
+# SPDX-License-Identifier: Apache-2.0
+"""vLLM server with KV cache NaN injection for debugging.
```
Comment (elvircrn) [inline serve_nan_debug.py]: TODO: Remove this file eventually.

File: test_nan.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
+import os
+os.environ["VLLM_ENABLE_V1_MULTIPROCESSING"] = "0"
```
Comment (elvircrn) [inline test_nan.py]: TODO: Also remove this one.

File: csrc/libtorch_stable/cache_kernels.cu [hunk 8]
```diff
@@ -421,9 +451,16 @@ __global__ void concat_and_cache_mla_kernel(
...
       if constexpr (kv_dt == Fp8KVCacheDataType::kAuto) {
         dst[dst_idx] = src[src_idx];
-      } else {
+      } else if constexpr (!kDebug) {
         dst[dst_idx] =
             fp8::scaled_convert<cache_t, scalar_t, kv_dt>(src[src_idx], *scale);
+      } else {
+        scalar_t sv = src[src_idx];
+        if (is_nan_bits(sv)) {
+          atomicAdd(debug_meta, 1);
```
Comment (elvircrn) [inline csrc/libtorch_stable/cache_kernels.cu]: TODO: Check if it's safe to do this without atomics at the end of the kernel.

File: vllm/envs.py (hunk 2 above)
Comment (elvircrn) [reply vllm/envs.py]: But I think that this being a debug flag only set by maintainers to help e2e debugging of vllm, makes it a candidate for it remaining an env flag, just like `VLLM_COMPUTE_NANS_IN_LOGITS`.

Lmk your thoughts on this.

File: tests/kernels/test_kv_cache_nan_e2e.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+                if sample.name == SAMPLE_NAME:
+                    prom_total += sample.value
+
+        assert prom_total == 0, (
+            f"No NaN injection but got {prom_total} NaN(s) in Prometheus"
+        )
+
+
+@pytest.mark.skipif(not CUDA_AVAILABLE, reason="Need CUDA device")
+class TestKVCacheNanMultiprocessing:
```
Comment (elvircrn) [inline tests/kernels/test_kv_cache_nan_e2e.py:134]: Claude-generated workaround.

File: csrc/libtorch_stable/cache_kernels.cu [hunk 10]
```diff
@@ -419,8 +423,10 @@ __global__ void concat_and_cache_mla_kernel(
   const int64_t block_idx = slot_idx / block_size;
   const int64_t block_offset = slot_idx % block_size;
 
+  int nan_count = 0;
+
   auto copy = [&](const scalar_t* __restrict__ src, cache_t* __restrict__ dst,
-                  int src_stride, int dst_stride, int size, int offset) {
+                  int src_stride, int size, int offset) {
     for (int i = threadIdx.x; i < size; i += blockDim.x) {
```
Comment (elvircrn) [inline csrc/libtorch_stable/cache_kernels.cu:430]: I know it's not a part of this PR, but `dst_stride` was unused so I just deleted it.
